client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,149 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { debounce, find } from 'lodash';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { toastr } from '@/services/ng';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+function search(term) {
+  // get recent
+  if (!term) {
+    return Query.recent().$promise
+      .then((results) => {
+        const filteredResults = results.filter(item => !item.is_draft); // filter out draft
+        return Promise.resolve(filteredResults);
+      });
+  }
+
+  // search by query
+  return Query.query({ q: term }).$promise
+    .then(({ results }) => Promise.resolve(results));
+}
+
+export function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  let isStaleSearch = false;
+  const debouncedSearch = debounce(_search, SEARCH_DEBOUNCE_DURATION);
+  const placeholder = 'Search a query by name';
+  const clearIcon = <i className="fa fa-times" onClick={() => setSelectedQuery(null)} />;
+  const spinIcon = <i className={cx('fa fa-spinner fa-pulse', { hidden: !searching })} />;
+
+  // set selected from prop
+  useEffect(() => {
+    if (props.selectedQuery) {
+      setSelectedQuery(props.selectedQuery);
+    }
+  }, [props.selectedQuery]);
+
+  // on search term changed, debounced
+  useEffect(() => {
+    // clear results, no search
+    if (searchTerm === null) {
+      setSearchResults(null);
+      return () => {};
+    }
+
+    // search
+    debouncedSearch(searchTerm);
+    return () => {
+      debouncedSearch.cancel();
+      isStaleSearch = true;
+    };
+  }, [searchTerm]);
+
+  // on query selected/cleared
+  useEffect(() => {
+    setSearchTerm(selectedQuery ? null : ''); // empty string forces recent fetch
+    props.onChange(selectedQuery);
+  }, [selectedQuery]);
+
+  function _search(term) {
+    setSearching(true);
+    search(term)
+      .then(rejectStale)
+      .then(setSearchResults)
+      .finally(() => {
+        setSearching(false);
```

Comment:
This will be called even if the results were rejected, but the new search might still be running... I think `setSearching` should be called as a `then` and not `finally`.

Reply:
Thought about that and it's true for rejection but not for any other error that might be thrown (e.g fetch error) - then it would keep going. I guess sth like this would cover it?
```js
search(term)
  .then(rejectStale)
  .then((results) => {
    setSearchResults(results);
    setSearching(false);
  })
  .catch((err) => {
    if (!(err instanceof StaleError)) {
      setSearching(false);
    }
  })
```

Reply:
I guess...

Reply:
That's exactly how I feel about it. I tend to go for a simple approach if the UX doesn't suffer that much. Whatever you choose.

Reply:
It can't stay the way it is now, because in some cases it will set it to not searching while in fact it does.

The other option is not to reject stale results (but just ignore them). I don't know what the implications of leaving pending promises though.

Reply:
Done.
https://github.com/getredash/redash/blob/cd9f1e0d92fe92d4c6d09176e3fd497f44bff4c6/client/app/components/QuerySelector.jsx#L70-L80